### PR TITLE
chore(pkcs#11): Add PKCS#11 documentation

### PR DESCRIPTION
**Pull request by MaciejTe**
chore(pkcs#11): Add PKCS#11 documentation

Changelog: None
Ticket: MEN-5760
Signed-off-by: Maciej Tomczuk <maciej.tomczuk@northern.tech>

# External Contributor Checklist

🚨 Please review the [guidelines for contributing](https://github.com/mendersoftware/mender/blob/master/CONTRIBUTING.md) to this repository.

- [x] Make sure that all commits follow the conventional commit [specification](https://www.github.com/mendersoftware/mendertesting/commitlint/grammar.md) for the Mender project.

The majority of our contributions are fixes, which means your commit should have
the form below:

```
fix: <SHORT DESCRIPTION OF FIX>

<OPTIONAL LONGER DESCRIPTION>

Changelog: <USER-FRIENDLY-CHANGE-DESCRIPTION> or <None>
Ticket: <TICKET NUMBER> or <None>
```

- [x] Make sure that all commits are signed with [`git --signoff`](https://git-scm.com/book/en/v2/Git-Tools-Signing-Your-Work). Also note that the signoff author must match the author of the commit.

### Description

Please describe your pull request.

Thank you!

### CODE REVIEW

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 1]
```diff
@@ -26,6 +26,59 @@ If the `ArtifactVerifyKey` option is set and Artifacts are not signed or the ver
...
+openssl_conf = openssl_init
+
+[openssl_init]
+engines = engine_section
+
+[engine_section]
+pkcs11 = pkcs11_section
+
+[pkcs11_section]
+dynamic_path = "/usr/lib/x86_64-linux-gnu/engines-1.1/pkcs11.so"
```
Comment (lluiscampos) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:56]: On latest Ubuntu, OpenSSL is on 3.0 version. For me the path is:
```suggestion
dynamic_path = "/usr/lib/x86_64-linux-gnu/engines-3/pkcs11.so"
```

So I this also deserves a note. Saying something like adjust `dynamic_path` according to your system. An easy way to find this file is to inspect the installed deb package with `dpkg -L libengine-pkcs11-openssl | grep /pkcs11.so`

Comment (kjaskiewiczz) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md:56]: something like [this](https://github.com/kjaskiewiczz/mender-docs/commit/42ab8aa0890f52d88e1991fa45169742b340bd56)?

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 2]
```diff
@@ -26,6 +26,59 @@ If the `ArtifactVerifyKey` option is set and Artifacts are not signed or the ver
...
+MODULE_PATH = "~/smtools-linux-x64/smpkcs11.so"
```
Comment (tranchitella) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:57]: This is a custom set up; smpkcs11 is a library provided by Digicert as part of the smtools package.
I don't think we should document it here.

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 3]
```diff
@@ -26,6 +26,59 @@ If the `ArtifactVerifyKey` option is set and Artifacts are not signed or the ver
...
+```
+
+!!! Engine shared object needs to be copied to proper location (in this case, SoftHSMv2 `~/smtools-linux-x64/smpkcs11.so` was used)
```
Comment (tranchitella) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:60]: This is not true, look at what we do to set up softHSM in our tests:
https://github.com/mendersoftware/integration/blob/master/tests/tests/test_mtls.py#L130

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 4]
```diff
@@ -26,6 +26,59 @@ If the `ArtifactVerifyKey` option is set and Artifacts are not signed or the ver
 
 ! If the Mender Client is configured to enable signature verification (through the `ArtifactVerifyKey` option), it will reject any unsigned Artifacts. This is necessary because otherwise an attacker could simply inject unsigned Artifacts to bypass the signature verification.
 
+## Supported cryptography standards
```
Comment (tranchitella) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:29]: I would move this section to the bottom of the document.
Using PKCS#11 is a somewhat more advanced use cases than using file-based keys.

**Review by tranchitella — CHANGES_REQUESTED** [review]

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 1 above)
Comment (lluiscampos) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md:56]: Yes :+1:

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 5]
```diff
@@ -124,3 +124,55 @@ For OS specific instructions on how to install and enable verification keys, vis
...
+pkcs11 = pkcs11_section
+[pkcs11_section]
+engine_id = pkcs11
+MODULE_PATH = /usr/lib/softhsm/libsofthsm2.so
+init = 0
+```
+
+You need to adjust MODULE_PATH according to your system.
+
+To make sure PKCS#11 engine is available for OpenSSL, issue following command:
```
Comment (tranchitella) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:160]:
```suggestion
To make sure PKCS#11 engine is available for OpenSSL, issue the following command:
```

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 6]
```diff
@@ -124,3 +124,55 @@ For OS specific instructions on how to install and enable verification keys, vis
...
+```bash
+user@mender:~$ openssl engine -t pkcs11
+(pkcs11) pkcs11 engine
+     [ available ]
+```
+
+After PKCS#11 engine is set up, artifact can be successfully signed with use of PKCS#11 URI:
```
Comment (tranchitella) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:167]:
```suggestion
Once the engine is set up, you can sign the Mender Artifact using a PKCS#11 URI:
```

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 7]
```diff
@@ -124,3 +124,55 @@ For OS specific instructions on how to install and enable verification keys, vis
...
+```bash
+user@mender:~$ ./mender-artifact sign --key-pkcs11 "pkcs11:object=device;type=private" artifact.mender
+```
+
+### Verifying
+
+PKCS#11 URI can also be used to verify signed artifacts:
```
Comment (tranchitella) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:174]:
```suggestion
You can also verify signed artifacts using the PKCS#11 interface:
```

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 8]
```diff
@@ -124,3 +124,55 @@ For OS specific instructions on how to install and enable verification keys, vis
...
+engines = engine_section
+[engine_section]
+pkcs11 = pkcs11_section
+[pkcs11_section]
+engine_id = pkcs11
+MODULE_PATH = /usr/lib/softhsm/libsofthsm2.so
+init = 0
+```
+
+You need to adjust MODULE_PATH according to your system.
```
Comment (tranchitella) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:158]:
```suggestion
You need to adjust MODULE_PATH according to your system and PKCS#11 interface.
```

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 4 above)
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 2 above)
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 9]
```diff
@@ -124,3 +124,55 @@ For OS specific instructions on how to install and enable verification keys, vis
...
+
+In order to sign an artifact using PKCS#11 interface, either hardware security modules (HSM), smart cards or Key Storage Provider (KSP) can be used.
+
+First, check if openssl and necessary libraries are installed:
+
+```bash
+sudo apt install openssl libengine-pkcs11-openssl
+```
+
+Next, proceed to OpenSSL configuration file:
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 5 above)
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 6 above)
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 7 above)
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 8 above)
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 10]
```diff
@@ -124,3 +124,55 @@ For OS specific instructions on how to install and enable verification keys, vis
...
+
+
+## Supported cryptography standards
+
+Mender supports following standards to interact with cryptographic tokens:
+* PKCS#11 (feature supported on Linux OS)
+
+### Signing
+
+In order to sign an artifact using PKCS#11 interface, either hardware security modules (HSM), smart cards or Key Storage Provider (KSP) can be used.
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 11]
```diff
@@ -124,3 +124,55 @@ For OS specific instructions on how to install and enable verification keys, vis
 ## Cloud Key Management
 
 It is possible to sign Artifacts using keys in Cloud Key Management, allowing developers to sign Mender Artifacts without ever accessing the private signing key. Currently the mender-artifact tool supports [Google Cloud Key Management](https://cloud.google.com/security-key-management?target=_blank). For more information, check the help screen for the `gcp-kms-key` option, available by running the command `mender-artifact sign --help`.
+
+
+## Supported cryptography standards
+
+Mender supports following standards to interact with cryptographic tokens:
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 3 above)
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 12]
```diff
@@ -124,3 +124,55 @@ For OS specific instructions on how to install and enable verification keys, vis
...
+
+
+## Supported cryptography standards
+
+Mender supports the following standards to interact with cryptographic tokens:
+* PKCS#11 (feature supported on Linux OS)
+
+### Signing
+
+You can use hardware security modules (HSMs), smart cards or Key Storage Provider (KSP) to sign an artifact using PKCS#11 interface.
```
Comment (tranchitella) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:136]:
```suggestion
You can use hardware security modules (HSMs), smart cards or Key Storage Provider (KSP) to sign an artifact using the PKCS#11 interface.
```

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 13]
```diff
@@ -124,3 +124,55 @@ For OS specific instructions on how to install and enable verification keys, vis
...
+
+First, check if openssl and necessary libraries are installed:
```
Comment (tranchitella) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:138]:
```suggestion
First, check if openssl and the necessary libraries are installed:
```

**Review by tranchitella — APPROVED** [review]
It looks good, I'd add two articles but I'm pre-approving.

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 12 above)
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 13 above)
Comment (MaciejTe) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md]: Done!

**Pull request merged**